internal/driver/nodeserver.go

```diff
@@ -86,15 +88,22 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	log, _, done := logger.GetLogger(ctx).WithMethod("NodePublishVolume")
 	defer done()
 
+	start := time.Now()
 	volumeID := req.GetVolumeId()
 	log.V(2).Info("Processing request", "volumeID", volumeID)
 
 	ns.mux.Lock()
 	defer ns.mux.Unlock()
 
+	success := metrics.SuccessTrue
+
 	// Validate the request object
 	log.V(4).Info("Validating request", "volumeID", volumeID)
 	if err := validateNodePublishVolumeRequest(ctx, req); err != nil {
+		success = metrics.SuccessFalse
+		// Record failure metric before returning
+		metrics.NodePublishTotal.WithLabelValues(success).Inc()
+		metrics.NodePublishDuration.WithLabelValues(success).Observe(time.Since(start).Seconds())
```

Comment:
Looks like these lines are being used multiple times, we should consider encapsulating it in some func or var?

Reply:
Yes, this makes sense. I have pushed changes to encapsulate this.